src/main/java/edu/usf/sas/pal/muser/manager/MuserMusicDataAnalysisManager.java

```diff
@@ -62,27 +77,100 @@ private void analyzeAllUsersMusicData() {
     private void processUserByIdForUiEvent(String userId) {
-        List<QueryDocumentSnapshot> userUiEventInfoById = new ArrayList<>(firebaseReader.getAllUserUiEventInfoById(userId));
+        List<QueryDocumentSnapshot> userUiEventInfoById = new ArrayList<>(firebaseReader.getUiEventInfoByUserId(userId));
         for (QueryDocumentSnapshot doc : userUiEventInfoById) {
-            MusicAnalysisModel mam = new MusicAnalysisModel(doc.toObject(MusicAnalysisModel.class), doc.getId(), userId);
-            musicAnalysisList.add(mam);
-            csvFileWriter.appendAllToCsV(musicAnalysisList);
-            musicAnalysisList.clear();
+            writeUserInfoToCSV(doc, userId);
+        }
+    }
+
+    //get data of Device info  collection based on user id
+    private void processUserByIdForDeviceInfo(String userId) {
+        List<QueryDocumentSnapshot> userDeviceInfoById = new ArrayList<>(firebaseReader.getDeviceInfoByUserId(userId));
+        for (QueryDocumentSnapshot doc : userDeviceInfoById) {
+            writeUserInfoToCSV(doc, userId);
         }
     }
+
+    //get data of Player Events collection based on date range and user id
+    private void processUserByDateRangeForPlayerEvent(String userId, long millisStart, long millisEnd) {
+        List<QueryDocumentSnapshot> userByDateRangeForPlayerEvent = new ArrayList<>(firebaseReader.getInfoByDateRangeForPlayerEventInfo(userId, millisStart, millisEnd));
+        for (QueryDocumentSnapshot doc : userByDateRangeForPlayerEvent) {
+            writeUserInfoToCSV(doc, userId);
+        }
+    }
+
+    //get data of Ui Events info collection on date range and user id
+    private void processUserByIdDateRangeForUiEvent(String userId, long millisStart, long millisEnd) {
+        List<QueryDocumentSnapshot> userByIdDateRangeForUiEvent = new ArrayList<>(firebaseReader.getInfoByDateRangeForUiEventInfo(userId, millisStart, millisEnd));
+        for (QueryDocumentSnapshot doc : userByIdDateRangeForUiEvent) {
+            writeUserInfoToCSV(doc, userId);
+        }
+    }
+
+    //get data of device info collection on date range and user id
+    private void processUserByIdDateRangeForDeviceInfo(String userId, long millisStart, long millisEnd) {
+        List<QueryDocumentSnapshot> userByIdDateRangeForUiEvent = new ArrayList<>(firebaseReader.getInfoByDateRangeForUiEventInfo(userId, millisStart, millisEnd));
+        for (QueryDocumentSnapshot doc : userByIdDateRangeForUiEvent) {
+            writeUserInfoToCSV(doc, userId);
+        }
+    }
+
+    //Write the Music Analysis Model to CSV
+    private void writeUserInfoToCSV(QueryDocumentSnapshot doc, String userId) {
+        MusicAnalysisModel mam = new MusicAnalysisModel(doc.toObject(MusicAnalysisModel.class), doc.getId(), userId);
+        musicAnalysisList.add(mam);
+        csvFileWriter.appendAllToCSV(musicAnalysisList);
+        musicAnalysisList.clear();
+    }
+
+    //below methodconverts start date to milliseconds
```

Comment:
Time is tricky in Java. There are several ways to do this, but I think the most reliable way is to use `LocalDate` instead:
https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html

This is a direct match to what is being passed in from the command-line - a date without any timezone or DST info.

Also, using `ZoneId.systemDefault()` would set the timezone and DST to the computer running the code. It may seem that's what we want now, but in the long term it makes this software unpredictable depending on where it's run. I'd prefer to pick a timezone and stick to it. We can hard-code to Eastern time if that makes the most sense for now, although typically you'd pick UTC time if the software is being used all over the world.

So the code would look something like:

```java
LocalDate startDate = LocalDate.parse(programOptions.getStartDate());
return startDate.atStartOfDay(ZoneId.of("America/New_York")).toEpochMilli();
```